src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -395,90 +410,97 @@ PayChanClaim::doApply()
-        return tecNO_TARGET;
+        return tecNO_ENTRY;
 
-    AccountID const src = (*slep)[sfAccount];
-    AccountID const dst = (*slep)[sfDestination];
-    AccountID const txAccount = ctx_.tx[sfAccount];
+    auto chanMembers = slep->getFieldArray (sfChannelMembers);
+    auto const& txClaims = ctx_.tx.getFieldArray (sfChannelClaims);
+    bool const thirdPartySubmission =
+        chanMembers[0].getAccountID (sfAccount) != ctx_.tx[sfAccount] &&
+        chanMembers[1].getAccountID (sfAccount) != ctx_.tx[sfAccount];
 
     auto const curExpiration = (*slep)[~sfExpiration];
+    if (curExpiration)
     {
-        auto const cancelAfter = (*slep)[~sfCancelAfter];
         auto const closeTime =
             ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
-        if ((cancelAfter && closeTime >= *cancelAfter) ||
-            (curExpiration && closeTime >= *curExpiration))
+        if (closeTime >= *curExpiration)
             return closeChannel (
                 slep, ctx_.view (), k.key, ctx_.app.journal ("View"));
-    }
 
-    if (txAccount != src && txAccount != dst)
+        // PayChanClaim without a claim does nothing if
+        // the channel has a pending expiration
+        if (txClaims.empty ())
+            return tecNO_PERMISSION;
+    }
+    // third party must have claim if the channel has not expired
+    else if (thirdPartySubmission && txClaims.empty ())
         return tecNO_PERMISSION;
 
-    if (ctx_.tx[~sfBalance])
+    bool submittedOwnClaim = false;
+
+    if (! txClaims.empty ())
     {
-        auto const chanBalance = slep->getFieldAmount (sfBalance).xrp ();
-        auto const chanFunds = slep->getFieldAmount (sfAmount).xrp ();
-        auto const reqBalance = ctx_.tx[sfBalance].xrp ();
+        std::vector<STAmount> amounts {
+            chanMembers[0][sfAmount],
+            chanMembers[1][sfAmount]};
 
-        if (txAccount == dst && !ctx_.tx[~sfSignature])
-            return temBAD_SIGNATURE;
+        std::vector<std::uint32_t> sequences {
+            chanMembers[0][sfSequence],
+            chanMembers[1][sfSequence]};
 
-        if (ctx_.tx[~sfSignature])
+        for (auto const& txClaim : txClaims)
         {
-            PublicKey const pk ((*slep)[sfPublicKey]);
-            if (ctx_.tx[sfPublicKey] != pk)
+            auto const i =
+                txClaim[sfPublicKey] == chanMembers[0][sfPublicKey] ? 0 : 1;
+
+            if (i == 1 && txClaim[sfPublicKey] != chanMembers[1][sfPublicKey])
                 return temBAD_SIGNER;
-        }
 
-        if (reqBalance > chanFunds)
-            return tecUNFUNDED_PAYMENT;
+            if (txClaim[sfSequence] < chanMembers[i][sfSequence])
+                return tefPAST_SEQ;
 
-        if (reqBalance <= chanBalance)
-            // nothing requested
-            return tecUNFUNDED_PAYMENT;
+            if (txClaim[sfSequence] == chanMembers[i][sfSequence] &&
+                    txClaim[sfAmount] <= chanMembers[i][sfAmount])
```

Comment:
It would me nice to have a rule that says "submit the last claims from each participant at this sequence level". That's an easy rule to document and explain. However, the `txClaim[sfAmount] <= chanMembers[i][sfAmount]` check prevents this (the equal case will trigger if that claim has already been submitted). How about a rule where the amount must be greater if signed by the other participant and must be greater or equal if self-signed? I think that prevents one party from keeping the channel open indefinitely while allowing the "last claims from each participant" rule. I haven't thought this through deeply, but wanted to bring it up.